api/src/server/auth.rs

```diff
@@ -141,32 +219,67 @@ pub async fn route_login(
 
     match request_token(oidc_client, query).await {
         Ok(Some((token, userinfo))) => {
-            let email: String = userinfo.email.clone().unwrap();
-            // Make the user
-            // TODO: handle setting username
-            // TODO: handle second login, right now it will error
-            let user: User = NewUser {
-                username: &email.chars().take(20).collect::<String>(),
+            let sub = userinfo.sub.clone().unwrap();
+
+            let existing_user_provider =
+                UserProvider::filter_by_sub_and_provider(&sub, &provider_name)
+                    .get_result::<UserProvider>(conn)
+                    .optional()
+                    .map_err(ResponseError::from)?;
+            match existing_user_provider {
+                Some(user_provider) => match user_provider.user_id {
+                    Some(user_id) => {
+                        // User already exists so normal login
+                        let user: User = users::table
+                            .filter(users::dsl::id.eq(user_id))
+                            .first(conn)
+                            .optional()
+                            .map_err(ResponseError::from)?
+                            .unwrap();
```

Comment:
The `optional` here is completely unnecessary, since you're just unwrapping the option anyway. You also can use `get_result` instead of `first`, so that it errors if there's more than one user (which obv is impossible but best practice is to use `get_result` whenever you expect exactly 1). So
`let user: User = users::table.filter(...).get_result().map_err(...)?;`